src/main.c

```diff
@@ -1,30 +1,145 @@
+    char *subnet_s;
+    char *allow_gateway_s;
+
+    struct target *target_list;
+
+    uint32_t *source_blacklist;
+    uint32_t *source_allowlist;
+
+    // DNS cache for hostnames in allow/exclude lists
+    dns_cache_t exclude_dns_cache;
+    dns_cache_t include_dns_cache;
+    int dns_cache_refresh_interval;
+
+    unsigned char eth_ip[4];
+    unsigned char gate_ip[4];
+
+    unsigned int  subnet;
+
+    unsigned char *buffer;
+    int sock_raw;
+    bool alive;
+    bool debug;
+} m;
+// Utility: resolve hostname to IPv4 address (returns 0 on success)
+int resolve_hostname(const char *hostname, uint32_t *out_ip) {
+	if (!hostname || !out_ip) return -1;
+	
+	struct addrinfo hints, *res = NULL;
+	memset(&hints, 0, sizeof(hints));
+	hints.ai_family = AF_INET;
+	hints.ai_socktype = SOCK_STREAM;
+	
+	int err = getaddrinfo(hostname, NULL, &hints, &res);
+	if (err != 0 || !res) return -1;
+	
+	struct sockaddr_in *addr = (struct sockaddr_in *)res->ai_addr;
+	if (!addr || res->ai_addrlen < sizeof(struct sockaddr_in)) {
+		freeaddrinfo(res);
+		return -1;
+	}
+	
+	*out_ip = addr->sin_addr.s_addr;
+	freeaddrinfo(res);
+	return 0;
+}
+
+// Non-blocking DNS cache refresh (only refresh if really needed and been a while)
+void refresh_dns_cache_if_needed(dns_cache_t *cache) {
+	if (!cache) return;
+	
+	time_t now = time(NULL);
+	if (cache->last_refresh == 0 || (now - cache->last_refresh) >= DEFAULT_DNS_CACHE_REFRESH_INTERVAL) {
+		DEBUG_PRINT("Refreshing DNS cache with %zu entries\n", cache->count);
+		
+		// Bounds check to prevent buffer overflow
+		size_t safe_count = (cache->count > MAX_SOURCE_LIST_ENTRIES) ? MAX_SOURCE_LIST_ENTRIES : cache->count;
+		
+		for (size_t i = 0; i < safe_count; ++i) {
+			// Additional safety checks
+			if (!cache->entries[i].entry || strlen(cache->entries[i].entry) == 0) continue;
+			
+			if (cache->entries[i].is_hostname) {
+				DEBUG_PRINT("Resolving hostname: %s\n", cache->entries[i].entry);
+				uint32_t resolved_ip;
+				if (resolve_hostname(cache->entries[i].entry, &resolved_ip) != 0) {
+					cache->entries[i].ip = 0; // Mark as unresolved
+					DEBUG_PRINT("Failed to resolve: %s\n", cache->entries[i].entry);
+				} else {
+					cache->entries[i].ip = resolved_ip;
+				}
+			} else {
+				// Parse as IPv4 string
+				uint8_t address[4];
+				int err = sscanf(cache->entries[i].entry, "%hhu.%hhu.%hhu.%hhu",
+					&address[0], &address[1], &address[2], &address[3]);
+				if (err == 4) {
+					// Safe conversion avoiding unaligned access
+					cache->entries[i].ip = ((uint32_t)address[0]) |
+					                       ((uint32_t)address[1] << 8) |
+					                       ((uint32_t)address[2] << 16) |
+					                       ((uint32_t)address[3] << 24);
+				} else {
+					cache->entries[i].ip = 0;
+				}
+			}
+		}
+		cache->last_refresh = now;
+	}
+}
+
+// Non-blocking DNS cache refresh (only refresh if really needed and been a while)
+void refresh_dns_cache_if_needed_nonblocking(dns_cache_t *cache) {
```

Comment:
Maybe I'm missing something. but how is this non-blocking? It doesn't spawn a separate thread.